torch/distributed/_spmd/parallel_mode.py

```diff
@@ -0,0 +1,101 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, List, Tuple
+
+import torch
+import torch.distributed as dist
+import torch.utils._pytree as pytree
+from torch.distributed._spmd.distribute import _convert_to_distributed, Schema
+from torch.distributed._tensor import DeviceMesh, Placement, Replicate, Shard
+
+from torch.fx import GraphModule
+
+
+class ParallelMode(ABC):
+    """Basic Parallel Mode interface."""
+
+    @abstractmethod
+    def expand(
+        self,
+        gm: GraphModule,
+        params_and_buffers: Dict[str, Any],
+        named_states: Dict[str, Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> GraphModule:
+        """expand a single device graph to a distributed graph."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def optimize(self, gm: GraphModule) -> GraphModule:
+        """optimize a distributed graph with a set of optimization passes"""
+        # TODO: add more necessary arguments to this interface.
+        raise NotImplementedError()
+
+    def configure_optimization_passes(self, _: List[Callable]) -> None:
+        """a way to configure optimization passes per parallel mode"""
+        raise NotImplementedError()
+
+
+class DTensorFallbackMode(ParallelMode):
+    """The DTensor fallback parallel mode. It's replicating the parameters
+    and shard the inputs to represent DDP like behavior, it's currently
+    a transitent mode before we move to the new data parallel expansion.
+    """
+
+    def __init__(self):
+        self._placements_override: Dict[int, List[Placement]] = {}
+        self._optimization_passes: List[Callable] = []
```

Comment:
Why not just take a callable and remove `optimize` method? This way it is more flexible for users to configure while not losing any connivence provided here.

Reply:
I think we need to hold a default set of optimization passes. where we explicitly register in this `__init__`. `optimize` calls the optimization passes in the order of this list one by one. Do you mean we have an optional callable argument in the init method, if not passed in, we use the default set of optimization, if passed in, we drop the default set and use the passed in callable?

That works for me :) We could get rid of `configure_optimization_passes` to favor of this!

The `optimize` method is merely a convenient callsite for the `compile` API to run the list of default optimization passes and lower to inductor to do "real" compilation, we can remove it for sure, but we would put this logic to the `compile` API itself.

Reply:
updated with suggestions